src/core/Debug.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  Disposable,
+  ExtensionMode,
+  Location,
+  TextEditorSelectionChangeEvent,
+  window,
+  workspace,
+} from "vscode";
+import { SyntaxNode, TreeCursor } from "web-tree-sitter";
+import { Graph } from "../typings/Types";
+
+const originalDebugLog = console.debug;
+const disabledDebugLog = () => {};
+
+console.debug = disabledDebugLog;
+
+export default class Debug {
+  getNodeAtLocation?: any;
+  disposable?: Disposable;
+
+  constructor(private graph: Graph) {
+    this.graph.extensionContext.subscriptions.push(this);
+
+    this.evaluateSetting = this.evaluateSetting.bind(this);
+
+    switch (this.graph.extensionContext.extensionMode) {
+      // Development mode. Always enable.
+      case ExtensionMode.Development:
+        this.enableDebugLog();
+        break;
+      // Test mode. Always disable.
+      case ExtensionMode.Test:
+        this.disableDebugLog();
+        break;
+      // Production mode. Enable based on user setting.
+      case ExtensionMode.Production:
+        this.evaluateSetting();
+        workspace.onDidChangeConfiguration(this.evaluateSetting);
+        break;
+    }
+  }
+
+  init(getNodeAtLocation: any) {
+    this.getNodeAtLocation = getNodeAtLocation;
+  }
+
+  dispose() {
+    if (this.disposable) {
+      this.disposable.dispose();
+    }
+  }
+
+  private enableDebugLog() {
+    console.debug = originalDebugLog;
+    this.disposable = window.onDidChangeTextEditorSelection(
+      this.logBranchTypes()
+    );
+  }
+
+  private disableDebugLog() {
+    console.debug = disabledDebugLog;
+    this.dispose();
+  }
+
+  private evaluateSetting() {
+    const debugEnabled = workspace
+      .getConfiguration("cursorless")
+      .get<boolean>("debug")!;
+    if (debugEnabled) {
+      this.enableDebugLog();
+    } else {
+      this.disableDebugLog();
+    }
+  }
+
+  private logBranchTypes() {
+    return (event: TextEditorSelectionChangeEvent) => {
```

Comment:
Why do you return a function here?  Why not just use the method itself (after binding it)?

Reply:
oh, that was how it was implemented before and I just copied it. I can clean it up